feat: add missing event type fields for advanced settings in api v2

dhairyashiil opened the pull request:
### Videos

_1. create event type:_


https://github.com/user-attachments/assets/bc9cf3eb-35d1-4bd6-a3b5-7ff7515622aa



_2. get single event type:_



https://github.com/user-attachments/assets/eb75bd82-87ea-488c-b719-3d2c934edea7



_3. get all event types:_


https://github.com/user-attachments/assets/03b5e09a-c03a-4f9f-ba14-6a839742f7d0



_4. update single event type:_


https://github.com/user-attachments/assets/63aea169-1329-4daa-8dfe-ff00bcc08bc3



_5. delete single event type:_


https://github.com/user-attachments/assets/e8e93e91-12da-4a01-85ce-41c9e68a208c



_6. Validation for interfaceLanguage:_


https://github.com/user-attachments/assets/bdbae79b-e73f-47b8-8517-e83089cf8f79


Note: Later I removed `customReplyToEmail` intentionally from this PR in [commit](https://github.com/calcom/cal.com/pull/25739/commits/68a847bb57360a1349218161a564f53e3c249b51). The website restricts this field to only allow the user's own verified emails via a dropdown, but implementing the same validation in the API requires additional logic. This will be addressed in a separate PR with proper email ownership validation.

dhairyashiil committed:
feat: api v2 add missing event type fields for advanced settings

Add 9 event type fields to API v2 that were available in tRPC but missing from the platform API:

- disableCancelling: disable cancelling for guests/organizer
- disableRescheduling: disable rescheduling for guests/organizer
- canSendCalVideoTranscriptionEmails: send Cal Video transcription emails
- autoTranslateInstantMeetingTitleEnabled: auto-translate instant meeting titles
- interfaceLanguage: preferred booking interface language
- allowReschedulingPastBookings: allow rescheduling past events
- allowReschedulingCancelledBookings: allow booking via reschedule link
- customReplyToEmail: custom reply-to email for confirmations
- showOptimizedSlots: optimize time slot arrangement

Updated output/input schemas, transformation services, and E2E tests.

dhairyashiil committed:
fix(api-v2): add proper OpenAPI types for nullable event type fields

Add explicit type and nullable properties to @ApiPropertyOptional decorators
for fields with `boolean | null` or `string | null` types. Without these,
Swagger was generating incorrect "type": "object" instead of the correct types.

Fixed fields:
- disableCancelling: type: Boolean, nullable: true
- disableRescheduling: type: Boolean, nullable: true
- interfaceLanguage: type: String, nullable: true
- allowReschedulingCancelledBookings: type: Boolean, nullable: true
- customReplyToEmail: type: String, nullable: true
- showOptimizedSlots: type: Boolean, nullable: true

pedroccastro reviewed (commented):
Hey @dhairyashiil! 👋 Nice PR! I've left a small suggestion about adding email validation for `customReplyToEmail` to prevent invalid values from causing email delivery issues downstream

dhairyashiil committed:
refactor: customReplyToEmail was intentionally excluded from this PR. The web UI
restricts this field to only allow the user's own verified emails via a
dropdown, but implementing the same validation in the API requires additional
business logic. This will be addressed in a separate PR with proper email
ownership validation.

pedroccastro commented:
> Actually, I removed `customReplyToEmail` from this PR 😅 because the website restricts this field to the user’s own verified emails through a dropdown. Implementing the same validation in the API would require additional logic. I might address this in a separate PR with proper email ownership validation.
> 
> thank you for the review @pedroccastro : )

Appreciate the context on `customReplyToEmail`, that makes sense! 👍
The tests are a bit verbose, but they’re consistent with the current pattern. I think a refactor to reduce duplication/complexity would be great for maintainability. Happy to track that as a follow-up PR since it might get out of scope here. Overall, looks good to me!

dhairyashiil committed:
feat(api-v2): add validation for interfaceLanguage field

Add @IsIn validation to interfaceLanguage field to only accept supported
locales. This ensures API v2 matches the web UI behavior where users can
only select from a predefined dropdown of supported languages.

Changes:
- Add SUPPORTED_LOCALES constant to @calcom/platform-constants
- Add @IsIn([...SUPPORTED_LOCALES]) validation to create/update DTOs
- Add E2E tests for invalid locale rejection (400 error)
- Add cross-reference comments in i18n.json and api.ts to keep in sync

supalarry commented on packages/platform/types/event-types/event-types_2024_06_14/inputs/create-event-type.input.ts:
```diff
@@ -512,6 +514,66 @@ export class BaseCreateEventTypeInput {
+  @DocsPropertyOptional({
+    description: "If true, guests and organizer can no longer reschedule the event.",
+  })
+  disableRescheduling?: boolean;
```
This PR https://github.com/calcom/cal.com/pull/25575 adds a new ability to not only toggle this setting on / off but also specify how many minutes before meeting we disable rescheduling.

So, to future proof this let's have `disableRescheduling` as object of type `DisableRescheduling` which right now only has 1 property to enable or disable this setting and then later on we can also add "when less than X minutes before the meeting setting".

I would recommend the same with "disableCancelling" just in case we add "when less than X minutes before the meeting setting" also there and because they are similar settings.

What do you think?

supalarry commented on packages/platform/types/event-types/event-types_2024_06_14/inputs/create-event-type.input.ts:
```diff
@@ -512,6 +514,66 @@ export class BaseCreateEventTypeInput {
+    description:
+      "When enabled, users will be able to create a new booking when trying to reschedule a cancelled booking.",
+  })
+  allowReschedulingCancelledBookings?: boolean;
```
I would note that this defaults to false.

supalarry reviewed (commented):
Nice PR. I have some comments below - let me know what you think.

dhairyashiil committed:
docs(api): add default values to event type input field documentation

Added default value documentation to @DocsPropertyOptional decorators
for the new event type fields in API v2 (2024_06_14):

- disableCancelling: false
- disableRescheduling: false
- canSendCalVideoTranscriptionEmails: true
- autoTranslateInstantMeetingTitleEnabled: false
- allowReschedulingPastBookings: false
- allowReschedulingCancelledBookings: false
- showOptimizedSlots: false

This improves API documentation by clearly communicating default
values to API consumers in the OpenAPI spec.

dhairyashiil committed:
Merge upstream/main into feat/add-event-type-advanced-settings-api-v2

supalarry commented:
> Can we handle both changes in separate follow-up PRs later? I need to focus on getting the companion app ready before Dec 15 😅
> 
> 1. Changing disableRescheduling and disableCancelling to objects
> 2. Moving canSendCalVideoTranscriptionEmails into calVideoSettings?
> 
> and added default values to all the new fields:
> 
> Field	Default
> `disableCancelling`	`false`
> `disableRescheduling`	`false`
> `canSendCalVideoTranscriptionEmails`	`true`
> `autoTranslateInstantMeetingTitleEnabled`	`false`
> `allowReschedulingPastBookings`	`false`
> `allowReschedulingCancelledBookings`	`false`
> `showOptimizedSlots`	`false`
> These defaults match the actual behavior on the website when creating a new event type.

API changes are hard to make later on - people might start using them and then we either need to:
1. introduce a new possible type for disableRescheduling thus making it more confusing for users and the boolean option forever will be considered the old way, but we will need to keep it because what if someone uses it.
2. deprecate canSendCalVideoTranscriptionEmails and hide it from docs but leave it in code because what if someone uses it

So when making an API it's better to just get it right the first time because later on we can't change api interface in a way that breaks it for existing users without creatinga new version of it.

I think 2 things above will take like 20 minutes to implement. We already have CalVideoSettings property so we just move property there and extract it from there and for disableRescheduling we create it as object instead - example is:
```
@ApiPropertyOptional({
    type: Routing,
    description:
      "Routing information from routing forms that determined the booking assignment. Both responseId and teamMemberIds are required if provided.",
    example: {
      responseId: 123,
      teamMemberIds: [101, 102],
    },
  })
  @IsOptional()
  @ValidateNested()
  @Type(() => Routing)
  routing?: Routing;
```

dhairyashiil committed:
Merge remote-tracking branch 'upstream/main' into feat/add-event-type-advanced-settings-api-v2

dhairyashiil committed:
feat(api-v2): refactor event type settings to object types for future extensibility

- Convert disableRescheduling from boolean to object with disabled and minutesBefore properties
- Convert disableCancelling from boolean to object with disabled property
- Move canSendCalVideoTranscriptionEmails into CalVideoSettings as sendTranscriptionEmails
- Remove autoTranslateInstantMeetingTitleEnabled (Enterprise-only feature)
- Add DisableRescheduling_2024_06_14 and DisableCancelling_2024_06_14 input/output types
- Add transformation methods for new object types in input and output services
- Update E2E tests for new API contract

BREAKING CHANGE: disableRescheduling and disableCancelling now accept objects instead of booleans

dhairyashiil committed:
fix(api-v2): prevent clearing calVideoSettings when only sendTranscriptionEmails is provided

Only include calVideoSettings in transformed output if it has properties,
avoiding unintentional reset of existing settings during updates.

dhairyashiil committed:
Merge branch 'main' into feat/add-event-type-advanced-settings-api-v2

supalarry reviewed (approved):
Amazing work as always Dhairyashil! Thank you for implementing the feedback 🚀

ThyMinimalDev reviewed (approved):
great work
